### ISSUE TITLE
Rule to detect changes to iterable object of loop

### ISSUE THREAD

**Issue by mimre25** [PyCQA/flake8-bugbear#445]
Recently I've ran into a classic bug in a code base, to illustrate a minimal example:

```py
some_list = [1,2,3]
for i in some_list:
  if i % 2 == 0:
    some_list.remove(i)
  print(i)
```
To nobodies surprise this prints only 1 and 2.

I was surprised that none of my linters caught this.

Anyway, would it be possible to add a rule that detects changes to the iterable that is being iterated?

I'm also more than happy to contribute :slightly_smiling_face:

**Comment by cooperlees** [comment]
Yeah, I think we've all been here.

This sounds good, but I think in general we should just detect if it is a mutable iterable when we see remove in a loop and suggest saving the indexs and then removing them in reverse? Or am I missing something here?

Anyways, in general this sounds good. Feel free to make a POC diff and we can comment on there.

**Comment by mimre25** [comment]
I've went with an approach that alerts for any mutations of the loop iterator (eg `.remove` or `del`).

PR is in #446

**Comment by mimre25** [comment]
FYI: While starting to implement this rule into `ruff` yesterday I might have found a bug in my initial implementation.

@cooperlees please hold off the release until I verify my suspicion and fix it ~ I'll check this out later today.

**Comment by mimre25** [comment]
Ok, I was seeing something, but that code wasn't used - I've opened #488 to remove it, to avoid future bugs :slightly_smiling_face:

**Comment by cooperlees** [comment]
This is released and we have dedicated improvements issue. So let's move there.

I yanked 24.1.15 + 24.1.15 due to bugs.
